doc/atdbg.1
.\"
.\" Source Version: 3.0
.\" Software Release #: LLNL-CODE-422942
.\"
.\" include cpyright.h
.\"

.TH ATDBG 1 "6 December 2004"
.SH NAME
atdbg \- attach a debugger to a named process
.SH SYNOPSIS
.B atdbg [-a] [-e <path>] [-h] [-p <pid>] [-q] [-r] [-v] [-w] [name]*
.SH DESCRIPTION
Present list of named jobs and attach a debugger to selected one.
The preferred debugger may be specified in order of preference with
the SC_DEBUGGER environment variable.  If SC_DEBUGGER is not defined,
it will check for one of the following debuggers in order: totalview,
gdb, or dbx.


.SH OPTIONS
.TP
.B \-a
Ignore user's ~/.ls-jobs-remove file to find all processes.
.TP
.B \-e
Full path to desired program.
.TP
.B \-h
A short summary help message.
.TP
.B \-p
Attach to the specified PID.
.TP
.B \-q
Quiet mode.
.TP
.B \-r
Do a call stack trace only.
.TP
.B \-v
Verbose mode.
.TP
.B \-w
Print the debugger being used and exit.


.SH EXAMPLES

   atdbg make
         This example will check all user jobs with the string make
         in the command and present each one to you with the option
         to attach a debugger it or not.

   atdbg -r -p 314 -e `which sx`
         This example will print the call stack of the executable sx
         running in process 314.

   atdbg -r -p 314
         This example will determine the executable running in
         process 314 and print the call stack of the current execution
         state.

   atdbg -w
         Prints out the full path of the debugger being used.

.SH ENVIRONMENT VARIABLES

SC_DEBUGGER  should be one or more "totalview", "gdb", or "dbx".
             For example, setenv SC_DEBUGGER gdb:totalview, will
             cause atdbg to look first for gdb then totalview.
.SH FILES

None.

.SH SEE ALSO

.B cleanup-jobs(1), ls-jobs(1), sign(1)

.SH BUGS
.TP 
.B None known.



manager/news
6/5/2012

For a long time the PACT Python bindings
were made using the Python Numeric package.  This has been
superceeded by the NumPy package.  When we build PACT
we look for both Numeric and NumPy now.  Because the two
packages have some differences we build a set of bindings for
each one that is available.  Since Numeric is the original package,
the bindings using it are in:

     /usr/local/apps/pactnew/gnu/lib/pythond.d/pact

and the new Numeric bindings are in:

     /usr/local/apps/pactnew/gnu/lib/pythond.d/pact_2

Each of those directories will be there if and only if the
Python package that it corresponds to is there.

The Chaos 5 machines appear to have only the NumPy package
and consequently only the pact_2 bindings.

To use the Numeric bindings you do:

    setenv PYTHONPATH  /usr/local/apps/pactnew/gnu/lib/pythond.d/pact

    and

    import pact.pdb as pdb

To use the NumPy bindings you do:

    setenv PYTHONPATH  /usr/local/apps/pactnew/gnu/lib/pythond.d/pact_2

    and

    import pact_2.pdb as pdb


1/26/2011

PDBLib supports C99 data types (version 26).

SCORE type manager expanded and made integral to
PACT in general.  The template utility was added to
generate functions for specified data types.  The
advantage over macros is debuggability and more
nearly first class behavior with respect to types.

More general polygon intersection handles non-simple,
non-convex polygons, multiple discreet polygon
intersections, and polygon unions as well.


11/10/2009

For developers, moved and consolidated "most"
#define'd constants into named enum's.  This allows the
compiler to do better error checking and helps in debugging
where you see symbolic names rather than numerical values.


For end users, there are many, many new features:

1) ULTRA now has a fill command to allow you to color
   the area under curves or fill closed curves.

2) Colors can be expressed as RGB values in many
   contexts.  I intend to get them all in time.
   An RGB color is specified as:

      #x7frrggbb

   where r, g, and b above are hex digits.

3) If you do the following in your .ultrarc

   ld mouse.scm

   you can can control the viewport with the
   mouse.  The functionality is more or less modelled
   after the Google map and other similar usages.

4) PDBView can now look at PDB files inside TAR
   files or AR archives.  This currently works only
   with a serial version of PDBView

5) PDBView can read mySQL and SQLite databases.
   Writing capabilities will come next.  This capability
   depends on the appropriate libraries and headers
   being installed on the platform where PACT is built.

6) A rudimentary ability to write PDB files which are
   plain text.  This will be fully developed in future
   releases.

01/10/2008

Collect PGS global attribute variables into struct and add
PG_get_attrs_glb, PG_set_attrs_glb, and PG_ptr_attrs_glb as the
sole accessors.  This keeps PGS state from conflicting with
applications or forcing them to change unnecessarily.

11/18/2007

Divide headers into public and private parts.  The publics parts
have the old names (e.g. score.h).  The private parts have the
old name with _int added (e.g. score_int.h).  Move most of the
_XX variables and functions into the private headers to hide them
from applications which should use standard accessors in the
public parts.


04/19/2006

Introduce PDB format version III.


08/13/2004

Introduce the hserve utility.


psh/atdbg.c
/*
 * ATDBG.C - attach a debugger to one of the named jobs 
 *
 * Source Version: 3.0
 * Software Release #: LLNL-CODE-422942
 *
 * #include "cpyright.h"
 *
 */

#include "common.h"
#include "libpsh.c"
#include "libtime.c"
#include "libasync.c"

#if 0
/* TV_CLI_1 refers to the CLI language used before about TV 8.11 */
#define HAVE_TV_CLI_1
#endif

/* TV_CLI_2 refers to the CLI language used from TV 8.11 to present */
#define HAVE_TV_CLI_2

#define ATTACH 10
#define TRACE  11

/* GDB actions */

#define QTHREAD 100
#define BTRACE  101

typedef struct s_atdbgdes atdbgdes;
typedef struct s_atproc atproc;
typedef struct s_dbgrsp dbgrsp;

struct s_atdbgdes
   {char os[BFLRG];
    char pname[BFLRG];
    char opt[BFLRG];
    char exe[BFLRG];
    char dbg[BFLRG];
    int qpid;                    /* PID from query */
    int spid;                    /* PID from command line */
    int mode;
    int verbose;
    int quiet;
    int whdbg;
    char **msgs;
    int nmsg;
    char logname[BFLRG];
    FILE *log;};

struct s_atproc
   {char name[BFLRG];
    int pid;};

struct s_dbgrsp
   {int action;
    char *prompt;
    char result[BFLRG];
    atdbgdes *st;};

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* CANDIDATE_PROC - return an array of candidate processes
 *                - derived from NAME
 */

static atproc *candidate_proc(atdbgdes *st, char *name)
   {int i, n;
    char pids[BFLRG];
    char *pl, *t;
    atproc *al;

    i  = 0;
    n  = 10;
    al = MAKE_N(atproc, n);
    if (al != NULL)

/* find the candidate process ids */
       {if ((IS_NULL(st->exe) == FALSE) && (st->spid > 0))
	   {al[i].pid = st->spid;
	    nstrncpy(al[i].name, BFLRG, st->exe, -1);
	    i++;}
	else
	   {nstrncpy(pids, BFLRG,
		     run(FALSE, "ls-jobs %s \"%s\" | sed '/atdbg /d' | sed '/sign /d' | awk '{print $2 \" \" $6}'",
			 st->opt, name), -1);
	    if (st->verbose == TRUE)
	       printf("%s\n", pids);

	    for (i = 0, pl = pids; TRUE; i++)
	        {t = strtok(pl, " \n");
		 if (t != NULL)
		    al[i].pid = atoi(t);
		 else
		    break;
		 pl = NULL;
		 t = strtok(pl, " \n");
		 if (t != NULL)
		    file_path(t, al[i].name, BFLRG);
/*		    nstrncpy(al[i].name, BFLRG,
			     run(FALSE, "rpath %s", t), -1); */
		 else
		    break;

		 if (i >= n-2)
		    {n += 10;
		     REMAKE(al, atproc, n);
		     if (al == NULL)
		        return(NULL);};};};

	 al[i].pid     = -1;
	 al[i].name[0] = '\0';};

    return(al);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* SELECT_PID - interactively select the appropriate process */

static atproc *select_pid(atdbgdes *st, atproc *al)
   {int i, ok;
    char resp[BFLRG];
    atproc *pal;

    ok = FALSE;
    for (i = 0, pal = al; pal->pid > 0; i++, pal++)
    
/* if no PID specified stop, present, and query */
	{if (st->spid > 0)
	    ok = (pal->pid == st->spid);

	 else
	    {printf("%s %s %d (y/n): ",
		    (st->mode == ATTACH) ? "attach" : "trace",
		    pal->name, pal->pid);

	     fgets(resp, BFLRG, stdin);
	     ok = (resp[0] == 'y');};

	 if (ok == TRUE)
	    {if (st->verbose == TRUE)
	        {printf("\n");
		 printf("Debugger:   %s\n", st->dbg);
		 printf("Executable: %s\n", pal->name);
		 printf("PID:        %d\n", pal->pid);};
	     break;};};

    if (ok == FALSE)
       pal = NULL;

    return(pal);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* SEND_MSG - prepare and send a message to PP */

static int send_msg(process *pp, int act, char *fmt, ...)
   {int nl;
    char msg[BFLRG];
    char *p;
    dbgrsp *gr;
    atdbgdes *st;

    VA_START(fmt);

    if (fmt == NULL)
       p = NULL;
    else
       {VSNPRINTF(msg, BFLRG, fmt);
	p  = msg;};

    VA_END;

    gr = (dbgrsp *) pp->a;
    st = gr->st;
    gr->action = act;

/* log the message sent */
    if ((st->log != NULL) && (p != NULL))
       fprintf(st->log, "SND: %s\n", p);

    nl = job_response(pp, 30000, p);

    return(nl);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* CACHE_RSP - put the message S in the cache for later transmission */

static void cache_rsp(atdbgdes *st, char *s)
   {

    if (st->msgs == NULL)
       {st->msgs = MAKE_N(char *, 100000);
        st->nmsg = 0;};

    if (st->msgs != NULL)
       st->msgs[st->nmsg++] = STRSAVE(s);

    return;}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* SEND_CACHED - now print the cached messages to stdout */

static void send_cached(atdbgdes *st)
   {int i, n;
    char r[BFLRG];
    char *s, **sa;

    n  = st->nmsg;
    sa = st->msgs;

/* print all the messages */
    for (i = 0; i < n; i++)
        {s = sa[i];

	 strclean(r, BFLRG, s, -1);

	 puts(r);

         FREE(s);};

/* reset */
    FREE(st->msgs);
    st->nmsg = 0;

    return;}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* RSP_ERROR - handle errors such as timeouts or poll rejects */

static int rsp_error(int fd, process *pp, char *t)
   {int ok;
    dbgrsp *gr;
    atdbgdes *st;

    gr = (dbgrsp *) pp->a;
    st = gr->st;

    if (LAST_CHAR(t) == '\n')
       LAST_CHAR(t) = '\0';

    if (st->log != NULL)
       fprintf(st->log, "ERR: while handling '%s'\n", t);

    ok = FALSE;

    return(ok);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* RSP_TOTALVIEW - handle lines from the totalview CLI */

static int rsp_totalview(int fd, process *pp, char *t)
   {int ok, nc;
    char p[BFLRG];
    char *s, *pa, *pb, *pr;
    dbgrsp *gr;
    atdbgdes *st;

    ok = FALSE;

    if ((pp != NULL) && (t != NULL))
       {gr = (dbgrsp *) pp->a;
	st = gr->st;
	pr = gr->prompt;
	nc = strlen(pr);

	if (LAST_CHAR(t) == '\n')
	   LAST_CHAR(t) = '\0';

	if (st->log != NULL)
	   fprintf(st->log, "RCV: %s\n", t);

	ok = (strncmp(t, pr, nc) == 0);

	if (ok == FALSE)
	   {switch (gr->action)
	       {case QTHREAD:
		     if (IS_NULL(t) == FALSE)
		        nstrncpy(gr->result, BFLRG, t, -1);
		     break;

		case BTRACE:
		     if (strstr(t, "PC=") != NULL)
		        {nstrncpy(p, BFLRG, t, -1);
			 pa = strstr(p, "PC=");
			 pb = strchr(p, '[');
			 if ((pa != NULL) && (pb != NULL))
			    memmove(pa, pb, strlen(pb)+1);
			 s = p;
			 s = subst(s, "PC=", " ", 1);
			 s = subst(s, "> ", "  ", 1);
			 s = subst(s, "#", ":", 1);
			 cache_rsp(st, s);
			 if (strstr(s, "SC_retrace_exe") != NULL)
			    cache_rsp(st, "         -+---------------------------+-");};
		     break;};};};

    return(ok);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* USE_TOTALVIEW - do the work with totalview CLI
 *               - NOTE: the function call arguments and values
 *               - are printed on separate lines which will require
 *               - some work to reassemble
 *               - for now just toss those lines
 */

static int use_totalview(atdbgdes *st, atproc *al)
   {int nl, rv, sig;
    char cmd[BFLRG], s[BFLRG];
    process *pp;
    dbgrsp gr;

    rv = TRUE;

    if (st->mode == TRACE)
       {gr.prompt = "d1.<> ";
	gr.st     = st;

	snprintf(cmd, BFLRG, "%scli", st->dbg);
	pp = job_launch(cmd, "a", NULL);
	if (pp != NULL)
	   {pp->accept = rsp_totalview;
	    pp->reject = rsp_error;
	    pp->a      = &gr;

	    send_msg(pp, 0, NULL);

#if defined(HAVE_TV_CLI_1)
/* TV versions < 8.11.0 */
	    send_msg(pp, 0, "dset LINES_PER_SCREEN 0");
	    send_msg(pp, 0, "dattach -no_attach_parallel %s %d",
		     al->name, al->pid);
	    send_msg(pp, 0, "dhalt");

	    send_msg(pp, QTHREAD, "dfocus p1 TV::thread get -all id");
	    FOREACH(thr, gr.result, " \n");
	       snprintf(s, BFLRG,
			"\n   ---------------- Thread %s ----------------\n", thr);
	       cache_rsp(st, s);
	       send_msg(pp, BTRACE, "dfocus %s {dwhere -a}", thr);
	    ENDFOR;

	    nl = send_msg(pp, 0, "exit -force");

#elif defined(HAVE_TV_CLI_2)

/* TV version >= 8.11.0 */
	    send_msg(pp, 0, "dattach %s %d", al->name, al->pid);
	    send_msg(pp, QTHREAD, "dfocus p1 TV::focus_threads");

	    FOREACH(thr, gr.result, " \n");
	       snprintf(s, BFLRG,
			"\n   ---------------- Thread %s ----------------\n", thr);
	       cache_rsp(st, s);
	       send_msg(pp, BTRACE, "dfocus %s {dhalt ; dwhere -a}", thr);
	    ENDFOR;

	    nl = send_msg(pp, 0, "exit -force");
#endif

	    sig = (nl == -1) ? SIGKILL : -1;
	    job_done(pp, sig);};}

    else
       run(FALSE, "%s %s -pid %d", st->dbg, al->name, al->pid);

    return(rv);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* RSP_GDB - handle lines from GDB */

static int rsp_gdb(int fd, process *pp, char *t)
   {int ok, nc;
    char p[BFLRG];
    char *s, *pr, *u;
    dbgrsp *gr;
    atdbgdes *st;

    gr = (dbgrsp *) pp->a;
    st = gr->st;
    pr = gr->prompt;
    nc = strlen(pr);

    if (LAST_CHAR(t) == '\n')
       LAST_CHAR(t) = '\0';

    if (st->log != NULL)
       fprintf(st->log, "RCV: %s\n", t);

    ok = (strncmp(t, pr, nc) == 0);

    if (ok == FALSE)
       {switch (gr->action)
	   {case QTHREAD:
	         u = strtok(t, " \t\n");
		 if ((u != NULL) && (*u == '*'))
		    u = strtok(NULL, " \t\n");
                 push_tok(gr->result, BFLRG, ' ', u);
	         break;

	    case BTRACE:
		 if (t[0] == '#')
		    {snprintf(p, BFLRG, "%s]", t+1);
		     s = subst(p, " (", "(", 1);
		     s = subst(s, " )", ")", 1);
		     s = subst(s, ") from ", ") [", 1);
		     s = subst(s, ") at ", ") [", 1);
		     s = subst(s, "../", "", -1);
		     s = subst(s, ", ", ",", -1);
		     s = subst(s, "<signal", "<----------signal", -1);
		     s = subst(s, "called>]", "called---------->", -1);
		     cache_rsp(st, s);
		     if (strstr(s, "SC_retrace_exe") != NULL)
		        cache_rsp(st, "         -+---------------------------+-");};
		 break;};};

    return(ok);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* USE_GDB - do the work with GDB */

static int use_gdb(atdbgdes *st, char *fname, atproc *al)
   {int rv, nl, sig;
    char cmd[BFLRG], s[BFLRG];
    process *pp;
    dbgrsp gr;

    rv = TRUE;

    if (st->mode == TRACE)
       {gr.prompt    = "(gdb) ";
	gr.result[0] = '\0';
	gr.st        = st;

	snprintf(cmd, BFLRG, "%s %s", st->dbg, al->name);
	pp = job_launch(cmd, "a", NULL);
	if (pp != NULL)
	   {pp->accept = rsp_gdb;
	    pp->reject = rsp_error;
	    pp->a      = &gr;

	    send_msg(pp, 0, NULL);
	    send_msg(pp, 0, "set width -1");
	    send_msg(pp, 0, "set height -1");
	    send_msg(pp, 0, "set confirm off");
	    send_msg(pp, 0, "attach %d", al->pid);
	    send_msg(pp, QTHREAD, "info threads");

	    FOREACH(thr, gr.result, " \n")
	       snprintf(s, BFLRG,
			"\n   ---------------- Thread %s ----------------\n", thr);
	       cache_rsp(st, s);
	       send_msg(pp, 0, "thread %s", thr);
	       send_msg(pp, BTRACE, "backtrace");
	    ENDFOR;

	    send_msg(pp, 0, "detach");

	    nl = send_msg(pp, 0, "quit");

	    sig = (nl == -1) ? SIGKILL : -1;
	    job_done(pp, sig);};}

    else
       run(FALSE, "xterm -sb -e %s %s %d", st->dbg, al->name, al->pid);

    return(rv);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* USE_DBX - do the work with DBX */

static int use_dbx(atdbgdes *st, char *fname, atproc *al)
   {int rv;
    char *s;
    FILE *fp;

    rv = TRUE;

    if (st->mode == TRACE)
       {fp = fopen_safe(fname, "w");
	if (fp != NULL)
	   {fprintf(fp, "attach %d\n", al->pid);
	    fprintf(fp, "where\n");
	    fprintf(fp, "quit\n");
	    fclose_safe(fp);

	    FOREACH(t, run(FALSE, "%s %s -c %s", st->dbg, al->name, fname), "\n")
	       s = t;
	       s = subst(s, ">", " ", 1);
	       cache_rsp(st, s);
	    ENDFOR;

	    unlink_safe(fname);};}

    else
       run(FALSE, "xterm -sb -e %s %s -pid %d", st->dbg, al->name, al->pid);

    return(rv);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* SESSION - work the session */

static int session(atdbgdes *st)
   {int i, n, rv, zip;
    char fname[BFLRG], names[BFLRG];
    char *name, *pl, *lst[33];
    atproc *al, *pal;
    FILE *lg;

    rv = TRUE;

    if (st->whdbg == TRUE)
       {printf("%s\n", st->dbg);
        exit(0);};

/* open the log file if requested */
    if (IS_NULL(st->logname) == FALSE)
       {snprintf(fname, BFLRG, "%s.%d", st->logname, (int) getpid());
	lg = fopen_safe(fname, "w");
	if (lg != NULL)
	   setbuf(lg, NULL);
	st->log = lg;};

    if (st->quiet == FALSE)
       printf("\n");

    if (st->mode == TRACE)
       snprintf(fname, BFLRG, "/tmp/dbg.%d", (int) getpid());

    if ((IS_NULL(st->pname) == TRUE) &&
        ((IS_NULL(st->exe) == FALSE) || (st->spid >= 0)))
       snprintf(st->pname, BFLRG, "%d", st->spid);

/* make an array of process names */
    nstrncpy(names, BFLRG, st->pname, -1);
    for (n = 0, pl = names; n < 32; n++, pl = NULL)
        {name = strtok(pl, " \n");
	 if (name == NULL)
	    break;
	 lst[n] = STRSAVE(name);};
    lst[n] = NULL;

/* check each name for matching processes */
    for (i = 0; i < n; i++)
        {name = lst[i];
	 zip  = 0;
	 al   = candidate_proc(st, name);

/* act on the list of process ids */
	 if (al == NULL)
            {zip++;
	     if (st->quiet == FALSE)
	        printf("No processes found under '%s'\n", name);}

	 else
	    {pal = select_pid(st, al);
	     if (pal != NULL)
	        {if (strstr(st->dbg, "totalview") != NULL)
		    rv = use_totalview(st, pal);

		 else if (strstr(st->dbg, "gdb") != NULL)
		    rv = use_gdb(st, fname, pal);

		 else if (strstr(st->dbg, "dbx") != NULL)
		    rv = use_dbx(st, fname, pal);

		 send_cached(st);

		 break;};};

	 FREE(al);
         FREE(lst[i]);};

    if (st->quiet == FALSE)
       printf("\n");

/* close the log file if requested */
    if (st->log != NULL)
       fclose_safe(st->log);

    return(rv);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* INIT - setup the session */

static int init(atdbgdes *st)
   {int rv;
    char lst[BFLRG], t[BFLRG];

    rv = TRUE;

    unamef(st->os, BFLRG, "s");
    nstrncpy(st->opt, BFLRG, "-q", -1);
    st->pname[0] = '\0';
    st->exe[0]   = '\0';
    st->dbg[0]   = '\0';
    st->spid     = -1;
    st->msgs     = NULL;
    st->nmsg     = 0;
    st->mode     = ATTACH;
    st->verbose  = FALSE;
    st->quiet    = FALSE;
    st->whdbg    = FALSE;

/* init logging */
    st->log = NULL;
    if (cdefenv("ATDBG_LOG") == TRUE)
       nstrncpy(st->logname, BFLRG, cgetenv(TRUE, "ATDBG_LOG"), -1);
    else
       st->logname[0] = '\0';

/* determine which debugger to use */
    if (cdefenv("SC_DEBUGGER") == TRUE)
       {nstrncpy(t, BFLRG, cgetenv(TRUE, "SC_DEBUGGER"), -1);
        nstrncpy(lst, BFLRG, subst(t, ":", " ", BFLRG), -1);}
    else
       nstrncpy(lst, BFLRG, "totalview gdb dbx", -1);

    FOREACH(i, lst, " \n")
       nstrncpy(t, BFLRG, cwhich(i), -1);
       if (file_executable(t) == TRUE)
          {nstrncpy(st->dbg, BFLRG, t, -1);
           break;};
    ENDFOR;

    if (IS_NULL(st->dbg) == TRUE)
       {printf("There is no recognized debugger - exiting\n");
        rv = FALSE;};

    return(rv);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* HELP - tell how to run atdbg */

static void help(void)
   {

    printf("\n");
    printf("Usage:  atdbg [-a] [-e <path>] [-h] [-l <file>] [-p <pid>] [-q] [-r] [-v] [-w] [name]*\n");
    printf("\n");
    printf("Present list of named jobs and attach debugger to selected one.\n");
    printf("\n");
    printf("        a    - search all jobs (i.e. ignore .ls-jobs-remove file)\n");
    printf("        e    - full path to the executable\n");
    printf("        h    - this help message\n");
    printf("        l    - log traffic to named file\n");
    printf("        p    - attach to specified PID\n");
    printf("        q    - quiet mode\n");
    printf("        r    - just show the current call stack\n");
    printf("        v    - verbose flag\n");
    printf("        w    - show which debugger is being used\n");
    printf("        name - executable name or command line pattern\n");
    printf("               used to identify processes\n");
    printf("\n");

    exit(1);

    return;}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* PROCESS_ARGS - process the command line arguments */

static int process_args(atdbgdes *st, int c, char **v)
   {int i;

    if (c == 0)
       help();

    for (i = 1; i < c; i++)
        {if (strcmp(v[i], "-a") == 0)
            {push_tok(st->opt, BFLRG, ' ', "-r");
	     push_tok(st->opt, BFLRG, ' ', "none-");}

	 else if (strcmp(v[i], "-e") == 0)
            nstrncpy(st->exe, BFLRG, v[++i], -1);

	 else if (strcmp(v[i], "-h") == 0)
            help();

	 else if (strcmp(v[i], "-l") == 0)
            nstrncpy(st->logname, BFLRG, v[++i], -1);

	 else if (strcmp(v[i], "-p") == 0)
            st->spid = atoi(v[++i]);

	 else if (strcmp(v[i], "-q") == 0)
            st->quiet = TRUE;

	 else if (strcmp(v[i], "-r") == 0)
            st->mode = TRACE;
/*
	 else if (strcmp(v[i], "-u") == 0)
            USER = v[++i];
*/
	 else if (strcmp(v[i], "-v") == 0)
	    st->verbose = TRUE;

	 else if (strcmp(v[i], "-w") == 0)
	    st->whdbg = TRUE;

	 else if (v[i][0] == '-')
	    continue;

	 else
	    push_tok(st->pname, BFLRG, ' ', v[i]);};

    return(i);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* MAIN - start it out here */

int main(int c, char **v)
   {int rv;
    char exe[BFLRG], bindir[BFLRG];
    atdbgdes state;

/* find the directory with atdbg */
    nstrncpy(exe, BFLRG, cwhich(v[0]), -1);
    nstrncpy(bindir, BFLRG, path_head(exe), -1);

/* locate the tools needed for subshells */
    build_path(NULL,
	       "sh", "xterm", "ls-jobs", "rpath",
	       NULL);

    push_path(P_PREPEND, lpath, bindir);

    rv = init(&state);
    if (rv == TRUE)
       {rv = process_args(&state, c, v);
	rv = session(&state);};

    return(rv);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

